### PR TITLE
Add support for multiple audio tracks

### LINKED ISSUE #5854: [STOP OPENING DUPLICATES] Support multiple audio tracks
### Checklist

- [x] I checked, but didn't find any duplicates (open OR closed) of this issue in the repo.
- [x] I have read the contribution guidelines given at https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md.
- [x] This issue contains only one feature request. I will open one issue for every feature I want to request.

#### Describe the feature you want
I noticed that some videos on this channel are played with non-English audio track: https://www.youtube.com/channel/UCBSs9x2KzSLhyyA9IKyt4YA (look for the "Multi-Audio Clip" ones)

The same video is played in English on mobile YouTube website. Changing the UI language (Settings -> Account -> Language) also changes the audio track language.

#### Is your feature request related to a problem? Please describe it
Yes. Some videos get played in a language I don't understand, while mobile YouTube web plays them in English by default (at least on my phone, tested on Firefox Klar/Focus with cookies cleared).

#### Additional context
See the channel link above.

#### How will you/everyone benefit from this feature?
This feature will let our users select which language they prefer and watch a video with corresponding audio track. It will also fix the problem I described above.

*Mod edit for search keywords: different other*

### PR DESCRIPTION
**Pull request by Theta-Dev**
#### What is it?
- [ ] Bugfix (user facing)
- [X] Feature (user facing)
- [ ] Codebase improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR
- Adds an audio language selector next to the video quality selector
- Default audio track is chosen by the configured content language. If no track matches, it defaults to English.
- New settings option: *Content* > *Prefer descriptive audio* (choose the descriptive audio track by default)

#### Before/After Screenshots/Screen Record

language_selector

#### Fixes the following issue(s)

- Fixes #5854

#### APK testing

- V1 (`208887d53`) https://github.com/TeamNewPipe/NewPipe/suites/11651391410/artifacts/605129298
- V2 (`ed06f559a`) https://github.com/TeamNewPipe/NewPipe/suites/11688854704/artifacts/607796221
- V3 https://github.com/TeamNewPipe/NewPipe/suites/11709640891/artifacts/609309922
- V4 (`39a5c8b`) https://github.com/TeamNewPipe/NewPipe/suites/11887030011/artifacts/621997850

#### Due diligence
- [X] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AudricV** [comment]
Thanks a lot! Here are some changes that I would suggest to do and that you need to do:

- For the audio settings you introduced, I would move the descriptive audio setting to `Video and audio`.
- For the language fallback you use, I would use the default locale instead of the English one, unless it creates some issues.
- For the audio track label, I am not sure if we should use the label set by the streaming service, as it creates localization issues. Track names provided by YouTube provides whether an audio track is an original or not, but that's only on recent videos, so we can't reliably extract this information on the label (and that's why I didn't added this property on the extractor, as I am not sure whether we should add it).
  
  Instead, I would use a string with the audio language translated in the app UI language and include the descriptive audio property if it is true with a string format (`%1s (%2s)` with `%1s` as the audio language and `%2s` as `descriptive`).

You need to fix the following behaviors:

- the background players (external and in-app) and video downloader ignore the settings you introduced;
- the downloader do not list the audio language for audio downloads;
- you should use the default audio language or the fallback language for video downloads at least, or better, if you want to do this in your PR and if other team members agree, add the audio selector below the video one: this would give flexibility to users. Someone else could always do the latter part :)
  
  This audio selector must be only present for demuxed streams and a message should be shown for muxed streams saying something like `An audio track should be already present in this stream`.
  The introduction of this message also raises the question to show all video streams and not merge video only and muxed video streams in the same list, as for instance, downloading a Mr Beast video in 720p with French audio track would not be possible in this case.

**Comment by AudricV** [comment]
> The menu suggested in following screenshot could be shown before playing a video on external player.

That's not possible, as most of video streams are demuxed, so no audio track is available in them.

> Not sure if it is within the scope of this issue.
> Maybe, Newpipe should do the same.

Yes it is out of the scope of the issue and is not possible except on downloads, but this will probably require a lot of work.

**Comment by AudricV** [comment]
Bug: the audio selection algorithm doesn't seem to take into account anymore the preferred audio format.

When selecting an audio track of latest Mr Beast's video for an external audio player with the M4A codec, I got the URL of a WEBM format (itag 251) where I should have got the itag 140.

Also, I don't know if that's intended or if you did changes since my comment about that, but no audio language is still shown in the downloader audio list.

**Comment by Theta-Dev** [comment]
Hi @AudricV,
I am still working on the downloader, so I have not pushed it yet.
I'll look into the codec selection issue.

**Comment by Theta-Dev** [comment]
Alright, I have everything implemented.

Do you know why the android tests time out?

**Review by Stypox — COMMENTED** [review]
> Default audio track is chosen by the configured content language. If no track matches, it defaults to English.

Shouldn't the default audio track chosen be the video's original lana

File: app/src/main/java/org/schabi/newpipe/download/DownloadDialog.java [hunk 1]
```diff
@@ -95,12 +97,14 @@ public class DownloadDialog extends DialogFragment
...
     @State
-    StreamSizeWrapper<AudioStream> wrappedAudioStreams;
-    @State
     StreamSizeWrapper<VideoStream> wrappedVideoStreams;
     @State
     StreamSizeWrapper<SubtitlesStream> wrappedSubtitleStreams;
     @State
+    AudioTracksWrapper wrappedAudioTracks;
+    @State
+    int selectedAudioStreamIndex;
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/download/DownloadDialog.java:106]:
```suggestion
    int selectedAudioTrackIndex;
```

File: app/src/main/java/org/schabi/newpipe/player/resolver/AudioPlaybackResolver.java [hunk 2]
```diff
@@ -59,34 +76,20 @@ public MediaSource resolve(@NonNull final StreamInfo info) {
         }
     }
 
-    /**
-     * Get a stream to be played as audio. If a service has no separate {@link AudioStream}s we
-     * use a video stream as audio source to support audio background playback.
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/resolver/AudioPlaybackResolver.java:64]: This comment shouldn't be lost

File: app/src/main/java/org/schabi/newpipe/player/ui/VideoPlayerUi.java [hunk 3]
```diff
@@ -1219,6 +1247,53 @@ public boolean onMenuItemClick(@NonNull final MenuItem menuItem) {
...
+        final int selectedStreamIndex = audioTrack.getSelectedAudioStreamIndex();
+        if (selectedStreamIndex == menuItemIndex || availableStreams.size() <= menuItemIndex) {
+            return;
+        }
+
+        player.saveStreamProgressState();
+        final String newAudioTrack = availableStreams.get(menuItemIndex).getAudioTrackId();
+        player.setRecovery();
+        player.setAudioTrack(newAudioTrack);
+        player.reloadPlayQueueManager();
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/player/ui/VideoPlayerUi.java:1292]: This part is also present in PlayQueueActivity. It's better to create a function on the player that combines these actions (maybe directly inside `setAudioTrack`, unless it's used in other places)

File: app/src/main/java/org/schabi/newpipe/util/ListHelper.java [hunk 4]
```diff
@@ -604,4 +667,148 @@ public static boolean isMeteredNetwork(@NonNull final Context context) {
...
+        }).thenComparing(bitrateComparator).thenComparingInt(
+                stream -> formatRanking.indexOf(stream.getFormat()));
+    }
+
+    /**
+     * Get a {@link Comparator} to compare {@link AudioStream}s by their tracks.
+     *
+     * <p>In this order:</p>
+     * <ol>
+     * <li>If {@code preferOriginalAudio}: is original audio</li>
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/util/ListHelper.java:720]:
```suggestion
     * <li>If {@code preferOriginalAudio}: use original audio</li>
```

File: app/src/main/java/org/schabi/newpipe/util/ListHelper.java [hunk 5]
```diff
@@ -604,4 +667,148 @@ public static boolean isMeteredNetwork(@NonNull final Context context) {
...
+        return getAudioTrackComparator(preferredLanguage, preferOriginalAudio,
+                preferDescriptiveAudio);
+    }
+
+    /**
+     * Get a {@link Comparator} to compare {@link AudioStream}s by their tracks.
+     *
+     * <p>In this order:</p>
+     * <ol>
+     * <li>If {@code preferOriginalAudio}: is original audio</li>
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/util/ListHelper.java:757]:
```suggestion
     * <li>If {@code preferOriginalAudio}: use original audio</li>
```

File: app/src/main/java/org/schabi/newpipe/util/StreamItemAdapter.java [hunk 6]
```diff
@@ -269,6 +269,10 @@ public static <X extends Stream> Single<Boolean> fetchSizeForWrapper(
                     .onErrorReturnItem(true);
         }
 
+        public void resetSizes() {
+            Arrays.fill(streamSizes, -2);
```
Comment (Stypox) [inline app/src/main/java/org/schabi/newpipe/util/StreamItemAdapter.java:273]: Maybe add a constant for `-2`

**Review by Stypox — CHANGES_REQUESTED** [review]
Things that should be tested before merging, on videos with various combinations of formats and languages. @Theta-Dev it would be great if you could provide a list of videos that have different features (e.g. no audio stream at all (i.e. only video streams with embedded audio), only original audio, descriptive audio, dubbed audio, all audio types in multiple languages, possibly not only from YouTube but from other platforms, too).
- downloading streams
- external player
- audio track menu both in the play queue activity and in the player

Thanks :-)

**Comment by Theta-Dev** [comment]
Videos for testing:
no external audio: any Peertube or CCC video
original+descriptive audio: https://www.youtube.com/watch?v=TjxC-evzxdk
original+dub: https://www.youtube.com/watch?v=_8W2LIfl5RE
original+dub+descriptive: https://www.youtube.com/watch?v=Kn56bMZ9OE8
multilanguage (no audio stream types): https://www.youtube.com/watch?v=tVWWp1PqDus&t=42
multilanguage (with audio stream types): https://www.youtube.com/watch?v=YLt73w6criQ

**Comment by Theta-Dev** [comment]
Okay, I applied the changes and updated the PR.

**Review by AudricV — CHANGES_REQUESTED** [review]
Thanks a lot!

Please address the following changes and try to avoid in the future that your IDE or yourself does formatting changes which make reviews harder.

File: app/src/main/java/org/schabi/newpipe/player/resolver/PlaybackResolver.java [hunk 7]
```diff
@@ -156,6 +156,11 @@ static String cacheKeyOf(final StreamInfo info, final AudioStream audioStream) {
             cacheKey.append(audioStream.getAverageBitrate());
         }
 
+        if (audioStream.getAudioTrackId() != null) {
+            cacheKey.append(" ");
+            cacheKey.append(audioStream.getAudioTrackId());
+        }
```
[the remark this answers is not in the document]
Comment (Theta-Dev) [reply app/src/main/java/org/schabi/newpipe/player/resolver/PlaybackResolver.java:162]: okay, I added the iso3-formatted language code in addition to the track ID

**Review by AudricV — APPROVED** [review]

**Comment by DebugKirby** [comment]
All the testing apk links are 404'd
How do I get this apk?

**Comment by ShareASmile** [comment]
As this code has been merged into dev branch. You can Try NewPipe Nightly builds:
https://github.com/TeamNewPipe/NewPipe-nightly/releases/tag/nightly-50

**Comment by TobiGr** [comment]
I found a regression in the latest nightly which was most likely introduced by this PR: #10180

**Comment by 15968133193-commits** [comment]
可以

**Pull request merged**